src/peergos/shared/user/SharedWithState.java

```diff
@@ -0,0 +1,150 @@
+        for (Map.Entry<String, Set<String>> e : readShares.entrySet()) {
+            if (e.getKey().equals(childName))
+                newReads.put(e.getKey(), new HashSet<>(e.getValue()));
+        }
+        Map<String, Set<String>> newWrites = new HashMap<>();
+        for (Map.Entry<String, Set<String>> e : writeShares.entrySet()) {
+            if (e.getKey().equals(childName))
+                newWrites.put(e.getKey(), new HashSet<>(e.getValue()));
+        }
+
+        return Optional.of(new SharedWithState(newReads, newWrites));
+    }
+
+    public SharedWithState add(SharedWithCache.Access access, String filename, Set<String> names) {
+        Map<String, Set<String>> newReads = new HashMap<>();
+        for (Map.Entry<String, Set<String>> e : readShares.entrySet()) {
+            newReads.put(e.getKey(), new HashSet<>(e.getValue()));
+        }
+        Map<String, Set<String>> newWrites = new HashMap<>();
+        for (Map.Entry<String, Set<String>> e : writeShares.entrySet()) {
+            newWrites.put(e.getKey(), new HashSet<>(e.getValue()));
+        }
+
+        if (access == SharedWithCache.Access.READ) {
+            newReads.putIfAbsent(filename, new HashSet<>());
+            newReads.get(filename).addAll(names);
+        } else if (access == SharedWithCache.Access.WRITE) {
+            newWrites.putIfAbsent(filename, new HashSet<>());
+            newWrites.get(filename).addAll(names);
+        }
+
+        return new SharedWithState(newReads, newWrites);
+    }
+
+    public SharedWithState remove(SharedWithCache.Access access, String filename, Set<String> names) {
+        Map<String, Set<String>> newReads = new HashMap<>();
+        for (Map.Entry<String, Set<String>> e : readShares.entrySet()) {
+            newReads.put(e.getKey(), new HashSet<>(e.getValue()));
+        }
+        Map<String, Set<String>> newWrites = new HashMap<>();
+        for (Map.Entry<String, Set<String>> e : writeShares.entrySet()) {
+            newWrites.put(e.getKey(), new HashSet<>(e.getValue()));
+        }
+
+        Set<String> val = access == SharedWithCache.Access.READ ? newReads.get(filename) : newWrites.get(filename);
+        if (val != null)
+            val.removeAll(names);
+
+        return new SharedWithState(newReads, newWrites);
+    }
+
+    public SharedWithState clear(String filename) {
+        Map<String, Set<String>> newReads = new HashMap<>();
+        for (Map.Entry<String, Set<String>> e : readShares.entrySet()) {
+            newReads.put(e.getKey(), new HashSet<>(e.getValue()));
+        }
+        Map<String, Set<String>> newWrites = new HashMap<>();
+        for (Map.Entry<String, Set<String>> e : writeShares.entrySet()) {
+            newWrites.put(e.getKey(), new HashSet<>(e.getValue()));
+        }
+
+        newReads.remove(filename);
+        newWrites.remove(filename);
+
+        return new SharedWithState(newReads, newWrites);
+    }
+
+    @JsMethod
+    public boolean isShared(String filename) {
```

Comment:
SharedWithState is internal. Not sure why this is a jsmethod.

Reply:
It's not internal. It's meant to be used by the UI directly. You retrieve it once for the current dir.

Reply:
I've added a helped method UserContext.getDirSharingState for this

Reply:
I think i'm missing something.
I can't see what class available to the UI returns an instance of SharedWithCache

Reply:
now up-to-date.